### PR TITLE
Added CheckParentingCycle, that check for any circular reference in the ancestors chain.

### LINKED ISSUE #358: Infinite hang when an added TreeNode forms a recursive hierarchy
* .NET Core Version: 3.0.100-preview-009812
* Have you experienced this same bug with .NET Framework? Yes

**Problem description:** A typo caused the wrong node to be added while constructing a hierarchy, causing a hang that was difficult to spot.

**Actual behavior:** Infinite hang in this loop:

https://github.com/dotnet/winforms/blob/fa4cdb0f50f8a64044a9d2baa192c53f6326f3e9/src/System.Windows.Forms/src/System/Windows/Forms/TreeNode.cs#L1575-L1576

```
System.Windows.Forms.dll!System.Windows.Forms.TreeNode.FindTreeView()
System.Windows.Forms.dll!System.Windows.Forms.TreeNode.TreeView.get()
System.Windows.Forms.dll!System.Windows.Forms.TreeNode.Realize(bool insertFirst)
System.Windows.Forms.dll!System.Windows.Forms.TreeNodeCollection.AddInternal(System.Windows.Forms.TreeNode node, int delta)
System.Windows.Forms.dll!System.Windows.Forms.TreeNodeCollection.Add(System.Windows.Forms.TreeNode node)	
```

**Expected behavior:** Just like the `Control.Controls` hierarchy, an exception should be thrown explaining that a TreeNode cannot be added as a descendant of itself.

**Minimal repro:**

```cs
var node = new TreeNode();
node.Nodes.Add(node); // Infinite hang
```

### PR DESCRIPTION
**Pull request by manuelvalenzuela**
Closes #358 

@manuelvalenzuela the closing comment has to go in this first message, not the PR header

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/TreeNode.cs [hunk 1]
```diff
@@ -1603,6 +1603,29 @@ private void EnsureStateImageValue()
...
 
+        /// <include file='doc\TreeNode.uex' path='docs/doc[@for="TreeNode.CheckParentingCycle"]/*' />
+        /// <devdoc>
+        ///     Check for any circular reference in the ancestors chain.
+        /// </devdoc>
+        internal void CheckParentingCycle(TreeNode candidateToAdd)
+        {
+            TreeNode node = this;
+
+            if (candidateToAdd == this)
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/TreeNode.cs]: Could you please refactor this method so that it has only a single `throw` statement?

Otherwise change looks good, I'm looking forward to the next commit with unit test for this particular exception. Thank you!

Comment (manuelvalenzuela) [reply src/System.Windows.Forms/src/System/Windows/Forms/TreeNode.cs]: Done!
thank you @Tanya-Solyanik!
Should I create a TreeNodeTests.cs file now?, I can see that is not created yet, otherwise, in which place should it be created?

Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/TreeNode.cs]: Hi @manuelvalenzuela ! Thank you for making this change. 
We messed up our naming convention for test files. Some files have Test suffix and some don't. The original intent was to name the file exactly as the class that is being tested, but append `Test` to the test class name. So tests for `TreeView` control are in `TreeViewTests` class in TreeView.cs file.  Here is that file:
https://github.com/dotnet/winforms/blob/master/src/System.Windows.Forms/tests/UnitTests/TreeView.cs
At this point it contains only a default test, please add your test tp this file.

File: src/System.Windows.Forms/tests/UnitTests/TreeView.cs [hunk 2]
```diff
@@ -20,5 +20,15 @@ public void TreeView_Constructor()
...
         }
+
+        [Fact]
+        public void TreeView_AddExistingNodeAsChild_ThrowsArgumentException()
+        {
+            var tv = new TreeView();
+            var tn1 = new TreeNode();
+            tv.Nodes.Add(tn1);
+
+            Assert.Throws<ArgumentException>(() => tn1.Nodes.Add(tn1));
```
Comment (hughbe) [inline src/System.Windows.Forms/tests/UnitTests/TreeView.cs]: add param name?

File: src/System.Windows.Forms/src/System/Windows/Forms/TreeNodeCollection.cs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Licensed to the .NET Foundation under one or more agreements.
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (hughbe) [inline src/System.Windows.Forms/src/System/Windows/Forms/TreeNodeCollection.cs]: what changed here?

File: src/System.Windows.Forms/src/System/Windows/Forms/TreeNode.cs [hunk 4]
```diff
@@ -1603,6 +1603,24 @@ private void EnsureStateImageValue()
             return total;
         }
 
+        /// <include file='doc\TreeNode.uex' path='docs/doc[@for="TreeNode.CheckParentingCycle"]/*' />
```
Comment (hughbe) [inline src/System.Windows.Forms/src/System/Windows/Forms/TreeNode.cs]: nit remove this

File: src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs [hunk 5]
```diff
@@ -973,6 +973,7 @@ private unsafe int SaveStreamToHandle(ref IntPtr handle, Stream stream)
             try
             {
                 var span = new Span<byte>(ptr.ToPointer(), size);
+                stream.Position = 0;
```
Comment (hughbe) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs]: is this an accident?

File: eng/Version.Details.xml [hunk 6]
```diff
@@ -1,43 +1,43 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Dependencies>
   <ProductDependencies>
-    <Dependency Name="Microsoft.Win32.Registry" Version="4.6.0-preview4.19115.1">
+    <Dependency Name="Microsoft.Win32.Registry" Version="4.6.0-preview4.19118.6">
```
Comment (hughbe) [inline eng/Version.Details.xml]: all this doesn't seem related

File: src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs (hunk 5 above)
Comment (manuelvalenzuela) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs]: Hi @hughbe, those are changes that were mixed from origin/master to my fork. Then the changes were mixed with mine and added back. You can see the original PR here: https://github.com/dotnet/winforms/pull/416

To keep my fork updated I execute this commands:
git fetch upstream
git rebase upstream/master

Do you suggest that I use another git command?
thank you!

File: eng/Version.Details.xml (hunk 6 above)
Comment (manuelvalenzuela) [reply eng/Version.Details.xml]: Same reason as before, mixed from master to my fork and added back: https://github.com/dotnet/winforms/pull/456

File: src/System.Windows.Forms/src/System/Windows/Forms/TreeNode.cs (hunk 4 above)
Comment (manuelvalenzuela) [reply src/System.Windows.Forms/src/System/Windows/Forms/TreeNode.cs]: removed, thank you!

File: src/System.Windows.Forms/src/System/Windows/Forms/TreeNodeCollection.cs (hunk 3 above)
Comment (manuelvalenzuela) [reply src/System.Windows.Forms/src/System/Windows/Forms/TreeNodeCollection.cs]: I can see that is a special character (\ufeff) related to the encoding of my PC, I will remove it.
Thank you

File: src/System.Windows.Forms/tests/UnitTests/TreeView.cs (hunk 2 above)
Comment (manuelvalenzuela) [reply src/System.Windows.Forms/tests/UnitTests/TreeView.cs]: Done!, thanks

File: src/System.Windows.Forms/tests/UnitTests/TreeView.cs [hunk 7]
```diff
@@ -20,5 +20,18 @@ public void TreeView_Constructor()
...
         }
+
+        public static Collections.Generic.IEnumerable<object[]> Add_ExistingNodeAsChild_TestData()
+        {
+            yield return new object[] { new TreeView(), new TreeNode() };
+        }
+        
+        [Theory]
+        [MemberData(nameof(Add_ExistingNodeAsChild_TestData))]
+        public void TreeView_AddExistingNodeAsChild_ThrowsArgumentException(TreeView treeView, TreeNode treeNode)
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/TreeView.cs]: Consider removing the theory from this case, I can't envision a different input dataset for this test case.

Comment (manuelvalenzuela) [reply src/System.Windows.Forms/tests/UnitTests/TreeView.cs]: Thanks @Tanya-Solyanik, I just did it.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs (hunk 5 above)
Comment (JuditRose) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs]: You can try merge the changes from dotnet:master on top of your changes instead of rebasing it. (It shouldn’t make any difference from this perspective, but maybe it fixes the mixups that wasn’t intentionally included in this PR). Alternatively you could try interactive rebasing and get rid of the unwanted commits (changes). Commits from Febr 19 are authored by others are seems like causing the mixups.

**Comment by manuelvalenzuela** [comment]
> Please update your fork so that unrelated changes don't show up as changed files. Perhaps you could reset your master to the commit before the last merge, and then re-run reset and cherry-pick your changes.
Done @Tanya-Solyanik! 
@JuditRose thank you so much for your advice!!!
